crates/aptos-rosetta/src/construction.rs

```diff
@@ -247,20 +241,64 @@ async fn construction_metadata(
     }
 
     let sequence_number = if let Some(sequence_number) = request.options.sequence_number {
-        sequence_number
+        sequence_number.0
     } else {
         // Retrieve the sequence number from the rest server if one wasn't provided
         response.inner().sequence_number
     };
 
+    // Determine max gas by simulation if it isn't provided
+    let max_gas_amount = if let Some(max_gas) = request.options.max_gas_amount {
+        max_gas.0
+    } else {
+        let transaction_factory = TransactionFactory::new(server_context.chain_id)
+            .with_gas_unit_price(1)
+            .with_max_gas_amount(u64::MAX);
+
+        let (txn_payload, sender) = request.options.internal_operation.payload()?;
+        let unsigned_transaction = transaction_factory
+            .payload(txn_payload)
+            .sender(sender)
+            .sequence_number(sequence_number)
+            .build();
+        let signed_transaction = SignedTransaction::new(
+            unsigned_transaction,
+            Ed25519PublicKey::try_from([0u8; 64].as_ref()).unwrap(),
+            Ed25519Signature::try_from([0u8; 64].as_ref()).unwrap(),
+        );
+
+        rest_client
+            .simulate_bcs(&signed_transaction)
+            .await?
+            .into_inner()
+            .info
+            .gas_used()
+    };
+
+    // Determine the gas price (either provided from upstream, or through estimation)
+    let gas_price_per_unit = if let Some(gas_price) = request.options.gas_price_per_unit {
+        gas_price.0
+    } else {
+        rest_client
+            .estimate_gas_price()
+            .await?
+            .into_inner()
+            .gas_estimate
+    };
+
+    let suggested_fee = Amount {
+        value: format!("-{}", gas_price_per_unit.saturating_mul(max_gas_amount)),
```

Comment:
Why isn't this just an actual value instead of a string? Rosetta API requirements?

Reply:
That's Rosetta :)